base/MAPL_SwathGridFactory.F90

```diff
@@ -0,0 +1,1405 @@
+!        if (any((this%lon_centers(2:im)-this%lon_centers(1:im-1))<0)) then
+!           where(this%lon_centers > 180) this%lon_centers=this%lon_centers-360
+!        end if
+!
+!
+!        v => file_metadata%get_coordinate_variable(lat_name,_RC)
+!        ptr => v%get_coordinate_data()
+!        _ASSERT(associated(ptr),'coordinate data not allocated')
+!        select type (ptr)
+!        type is (real(kind=REAL64))
+!           this%lat_centers = ptr
+!        type is (real(kind=REAL32))
+!           this%lat_centers = ptr
+!        class default
+!           _FAIL('unsupported type of data; must be REAL32 or REAL64')
+!        end select
+!
+!
+!        ! Check: is this a "mis-specified" pole-centered grid?
+!        if (size(this%lat_centers) >= 4) then
+!           ! Assume lbound=1 and ubound=size for now
+!           i_min = 1 !lbound(this%lat_centers)
+!           i_max = size(this%lat_centers) !ubound(this%lat_centers)
+!           d_lat = (this%lat_centers(i_max-1) - this%lat_centers(i_min+1))/&
+!                    (size(this%lat_centers)-3)
+!           is_valid = .True.
+!           ! Check: is this a regular grid (i.e. constant spacing away from the poles)?
+!           do i=(i_min+1),(i_max-2)
+!              d_lat_temp = this%lat_centers(i+1) - this%lat_centers(i)
+!              is_valid = (is_valid.and.(abs((d_lat_temp/d_lat)-1.0) < 1.0e-5))
+!              if (.not. is_valid) then
+!                 exit
+!              end if
+!           end do
+!           if (is_valid) then
+!              ! Should the southernmost point actually be at the pole?
+!              extrap_lat = this%lat_centers(i_min+1) - d_lat
+!              if (extrap_lat <= ((d_lat/20.0)-90.0)) then
+!                 this%lat_centers(i_min) = -90.0
+!              end if
+!              ! Should the northernmost point actually be at the pole?
+!              extrap_lat = this%lat_centers(i_max-1) + d_lat
+!              if (extrap_lat >= (90.0-(d_lat/20.0))) then
+!                 this%lat_centers(i_max) =  90.0
+!              end if
+!           end if
+!        end if
+!
+!        if (use_file_coords) then
+!           this%lon_centers = MAPL_DEGREES_TO_RADIANS_R8 * this%lon_centers
+!           this%lat_centers = MAPL_DEGREES_TO_RADIANS_R8 * this%lat_centers
+!        end if
+!         
+    end associate
+
+    call this%make_arbitrary_decomposition(this%nx, this%ny, _RC)
+
+    ! Determine IMS and JMS with constraint for ESMF that each DE has at least an extent
+    ! of 2.  Required for ESMF_FieldRegrid().
+    allocate(this%ims(0:this%nx-1))
+    allocate(this%jms(0:this%ny-1))
+    call MAPL_DecomposeDim(this%im_world, this%ims, this%nx, min_DE_extent=2)
+    call MAPL_DecomposeDim(this%jm_world, this%jms, this%ny, min_DE_extent=2)
+
+    call this%check_and_fill_consistency(_RC)
+
+    _RETURN(_SUCCESS)
+
+   end subroutine initialize_from_file_metadata
+
+
+
+   subroutine initialize_from_config_with_prefix(this, config, prefix, unusable, rc)
```

Comment:
This procedure might be "too long".  But maybe I'll feel differently after the commented sections are removed.

Worst case just need to add a few helper procedures that encapsulate pieces of this.